### PR TITLE
Use AppVeyor's included Miniconda

### PR DESCRIPTION
**Pull request by jakirkham**
Fixes https://github.com/conda-forge/conda-smithy/issues/351
Closes https://github.com/conda-forge/conda-smithy/issues/176
Closes https://github.com/conda-forge/setuptools-feedstock/pull/26

This PR changes the feedstock behavior to use AppVeyor's included copy of Miniconda instead of downloading our own. It does this by borrowing similar changes used in `staged-recipes` thanks to PR ( https://github.com/conda-forge/staged-recipes/pull/982 ) and applies them to the feedstocks. Hopefully this should avoid some issues that we are experiencing with Python 2.7 builds on AppVeyor where we are seeing long startup times. Also it should save some time by not needing to reinstall everything from scratch each time.

Testing of this is occurring in PR ( https://github.com/conda-forge/setuptools-feedstock/pull/26 ).

### CODE REVIEW

**Comment by MSeifert04** [comment]
I had some problems with this change:

a) it doesn't fix the memoryerror (see https://ci.appveyor.com/project/MSeifert04/iteration-utilities-feedstock/build/job/07sbeulj118lidht)
b) it complains about conda not being a valid command.

b) is easily fixed by including:
 ```
@@ -1,13 +1,10 @@
 # This file was automatically generated by conda-smithy. To update a component of this
 # file, make changes to conda-forge.yml and/or recipe/meta.yaml, and run
 # "conda smithy rerender".

 environment:
-
-  CONDA_INSTALL_LOCN: "C:\\conda"
-
   # SDK v7.0 MSVC Express 2008's SetEnv.cmd script will fail if the
   # /E:ON and /V:ON options are not enabled in the batch script intepreter
   # See: http://stackoverflow.com/a/13751649/163740
   CMD_IN_ENV: "cmd /E:ON /V:ON /C obvci_appveyor_python_build_env.cmd"
 
@@ -19,25 +16,31 @@ environment:
     secure: MP4hZYylDyUWEsrt3u3cod2sbFeRwUziH02mvQOdbjsTO/l1yIxDkP/76rSIjcGC
 
   matrix:
     - TARGET_ARCH: x86
       CONDA_PY: 27
+      CONDA_INSTALL_LOCN: "C:\\Miniconda"
 
     - TARGET_ARCH: x64
       CONDA_PY: 27
+      CONDA_INSTALL_LOCN: "C:\\Miniconda-x64"
 
     - TARGET_ARCH: x86
       CONDA_PY: 34
+      CONDA_INSTALL_LOCN: "C:\\Miniconda3"
 
     - TARGET_ARCH: x64
       CONDA_PY: 34
+      CONDA_INSTALL_LOCN: "C:\\Miniconda3-x64"
 
     - TARGET_ARCH: x86
       CONDA_PY: 35
+      CONDA_INSTALL_LOCN: "C:\\Miniconda35"
 
     - TARGET_ARCH: x64
       CONDA_PY: 35
+      CONDA_INSTALL_LOCN: "C:\\Miniconda35-x64"
 
 
 # We always use a 64-bit machine, but can build x86 distributions
 # with the TARGET_ARCH variable.
 platform:
@@ -54,12 +57,10 @@ install:
         Where-Object pullRequestId -eq $env:APPVEYOR_PULL_REQUEST_NUMBER)[0].buildNumber) { `
           throw "There are newer queued builds for this pull request, failing early." }
 
     # Cywing's git breaks conda-build. (See https://github.com/conda-forge/conda-smithy-feedstock/pull/2.)
     - cmd: rmdir C:\cygwin /s /q
-    - appveyor DownloadFile "https://raw.githubusercontent.com/conda-forge/conda-smithy/e976c7e84bb3c4846e7562afd1a42c4b535b51f5/bootstrap-obvious-ci-and-miniconda.py"
-    - cmd: python bootstrap-obvious-ci-and-miniconda.py %CONDA_INSTALL_LOCN% %TARGET_ARCH% %CONDA_PY:~0,1% --without-obvci
 
     # Add a hack to switch to `conda` version `4.1.12` before activating.
     # This is required to handle a long path activation issue.
     # Please see PR ( https://github.com/conda/conda/pull/3349 ).
     - cmd: set "OLDPATH=%PATH%"
```

**Comment by jakirkham** [comment]
Wouldn't say this change is ready to go anywhere yet. Hence why I hadn't pinged you about it. 😉

**Comment by jakirkham** [comment]
Alright this seems to be working on my fork now. Please take a look and let me know your thoughts. This uses the included Miniconda on AppVeyor. It also fixes a crucial bug.

xref: https://ci.appveyor.com/project/jakirkham/setuptools-feedstock/build/1.0.5

Side note: AFAICT it doesn't take notably longer to upgrade `conda` compared to what it was doing before.

cc @conda-forge/core @MSeifert04 @janschulz

**Comment by MSeifert04** [comment]
What was the crucial bug? 😄 

The changes actually speed up my builds (2:30 -> 1:50 on py3) and the 27 builds didn't fail with the memory-error.

I don't know why you originally used the development channel for obvious-ci but the used version changed from [0.5.2.dev2](https://ci.appveyor.com/project/conda-forge/geopandas-feedstock/build/1.0.51/job/1ggnx8velb7i1dr5#L69) to [0.6.1](https://ci.appveyor.com/project/jakirkham/setuptools-feedstock/build/1.0.5/job/0694g06ty4ka1q5i#L70) . Don't know if there was any reason for this pinning, just wanted to make sure that this was intentional.

File: conda_smithy/templates/appveyor.yml.tmpl [hunk 1]
```diff
@@ -46,25 +44,34 @@ install:
...
+    - cmd: if "%CONDA_PY%" == "27" if  "%TARGET_ARCH%" == "x86" set "CONDA_INSTALL_LOCN=C:\\Miniconda"
+    - cmd: if "%CONDA_PY%" == "27" if  "%TARGET_ARCH%" == "x64" set "CONDA_INSTALL_LOCN=C:\\Miniconda-x64"
+    - cmd: if "%CONDA_PY%" == "34" if  "%TARGET_ARCH%" == "x86" set "CONDA_INSTALL_LOCN=C:\\Miniconda3"
+    - cmd: if "%CONDA_PY%" == "34" if  "%TARGET_ARCH%" == "x64" set "CONDA_INSTALL_LOCN=C:\\Miniconda3-x64"
+    - cmd: if "%CONDA_PY%" == "35" if  "%TARGET_ARCH%" == "x86" set "CONDA_INSTALL_LOCN=C:\\Miniconda35"
+    - cmd: if "%CONDA_PY%" == "35" if  "%TARGET_ARCH%" == "x64" set "CONDA_INSTALL_LOCN=C:\\Miniconda35-x64"
+
+    # Add our channels.
+    - cmd: set "OLDPATH=%PATH%"
+    - cmd: set "PATH=%CONDA_INSTALL_LOCN%\\Scripts;%CONDA_INSTALL_LOCN%\\Library\\bin;%PATH%"
```
Comment (MSeifert04) [inline conda_smithy/templates/appveyor.yml.tmpl]: This line and the one above belong to the `# Use AppVeyor's preinstalled Miniconda.`-section, right?

Comment (jakirkham) [inline conda_smithy/templates/appveyor.yml.tmpl]: Actually these lines *technically* belong below with the hack. However, pulling things from `conda-forge` seems to fix the bug we were experiencing. So it had to go up here.

**Comment by jakirkham** [comment]
> I don't know why you originally used the development channel for obvious-ci but the used version changed from [0.5.2.dev2](https://ci.appveyor.com/project/conda-forge/geopandas-feedstock/build/1.0.51/job/1ggnx8velb7i1dr5#L69) to [0.6.1](https://ci.appveyor.com/project/jakirkham/setuptools-feedstock/build/1.0.5/job/0694g06ty4ka1q5i#L70) . Don't know if there was any reason for this pinning, just wanted to make sure that this was intentional.

Not entirely sure of the differences myself as `0.5.2.dev2` does not exist as a tagged release. At least between `0.5.1` and `0.6.1` the [changes]( https://github.com/pelson/Obvious-CI/compare/v0.5.1...v0.6.1 ) at the repo, appear to be simple things like relicensing and some Python 2/3 compatibility fixes. That being said, we already have been upgrading to `0.6.1` at staged-recipes for some time and that has worked fine. So I don't think that there is any good reason to use the older version in the feedstocks.

**Comment by jakirkham** [comment]
> I don't know why you originally used the development channel for obvious-ci but the used version changed from [0.5.2.dev2](https://ci.appveyor.com/project/conda-forge/geopandas-feedstock/build/1.0.51/job/1ggnx8velb7i1dr5#L69) to [0.6.1](https://ci.appveyor.com/project/jakirkham/setuptools-feedstock/build/1.0.5/job/0694g06ty4ka1q5i#L70) . Don't know if there was any reason for this pinning, just wanted to make sure that this was intentional.

Not entirely sure of the differences myself as `0.5.2.dev2` does not exist as a tagged release. At least between `0.5.1` and `0.6.1` the [changes]( https://github.com/pelson/Obvious-CI/compare/v0.5.1...v0.6.1 ) at the repo, appear to be simple things like relicensing and some Python 2/3 compatibility fixes.

That being said, we already have been upgrading to `0.6.1` at staged-recipes for some time and that has worked fine. So I don't think that there is any good reason to use the older version in the feedstocks nor why it should not come from conda-forge.

**Comment by jakirkham** [comment]
Cleaned up a bit more to use proper matrix environment variables for specifying the Miniconda installer location to use.

**Comment by jakirkham** [comment]
> What was the crucial bug? 😄

My suspicion is it is some package from `defaults` channel that acts up as getting everything possible from the `conda-forge` channel does not run into this issue. Though it could just as easily be a bug in a later version of `conda`.

AFAICT Switching to the pre-included Miniconda doesn't seem to make a difference.

**Comment by jakirkham** [comment]
Going to go ahead and get this out there as the AppVeyor startup lag is pretty painful. Will release as `v1.5.0`.

File: bootstrap-obvious-ci-and-miniconda.py [hunk 2]
```diff
@@ -1,103 +0,0 @@
-#!/usr/bin/env python
```
Comment (jakirkham) [inline bootstrap-obvious-ci-and-miniconda.py]: All the recent re-renderings point to the commit where this is and it is still available. Have just now rechecked.

https://raw.githubusercontent.com/conda-forge/conda-smithy/e976c7e84bb3c4846e7562afd1a42c4b535b51f5/bootstrap-obvious-ci-and-miniconda.py

Comment (ocefpaf) [inline bootstrap-obvious-ci-and-miniconda.py]: Just don't remove it. We need a `master` version in many projects. Do not be obtuse here. There are many outside `conda-forge` who use that.

Comment (jakirkham) [inline bootstrap-obvious-ci-and-miniconda.py]: Not trying to be obtuse. Only pointing out it is still available and that we had tried to protect against problems in conda-forge. In any event, PR ( https://github.com/conda-forge/conda-smithy/pull/376 ) will add it back. Feel free to merge.

Comment (jakirkham) [inline bootstrap-obvious-ci-and-miniconda.py]: Done in PR ( https://github.com/conda-forge/conda-smithy/pull/377 ).

Comment (pelson) [inline bootstrap-obvious-ci-and-miniconda.py]: @ocefpaf - I'm trying to catch up with some notifications here, and have no idea if there is more context to this conversation (if there is, I'm sorry), but I'd just like to ask that you try to reduce the bluntness of comments like this. I fully approve of directness (i.e. we don't need to beat around the bush), but it is just as easy to be direct *and polite*...

> Why did you removed this!!!!! This is need by many outside of conda-forge!!!!

Could just have easily been written:

> This change will break things for users outside of conda-forge. Would you mind adding it back please?

---------

I understand frustration when on change results in breakage elsewhere, but in general the world is a nicer place when we tone down the aggression (take that as a political statement if you like 😜 ) . No response needed here, I just want to raise the matter for you to consider in future comments.

Thank you.

**Pull request merged**